File: bot/feeds/espn.py
"""ESPN scoreboard client (keyless). Carries sportsbook odds blocks per event."""
from __future__ import annotations

from .base import LiveEvent, MarketPrice, OutcomePrice, Source
from .http import get_json

BASE = "https://site.api.espn.com/apis/site/v2/sports"

# ESPN's edge WAF blocks browser-like UAs; a neutral client UA is accepted.
ESPN_UA = "live-odds-bot/1.0 (+https://github.com/)"


# sport key -> ESPN league path
SPORT_PATHS = {
    "soccer-epl": "soccer/eng.1",
    "soccer-laliga": "soccer/esp.1",
    "soccer-bundesliga": "soccer/ger.1",
    "soccer-seriea": "soccer/ita.1",
    "soccer-ligue1": "soccer/fra.1",
    "soccer-ucl": "soccer/uefa.champions",
    "nba": "basketball/nba",
    "nfl": "football/nfl",
    "mlb": "baseball/mlb",
    "nhl": "hockey/nhl",
    "tennis-atp": "tennis/atp",
}


def _american_to_decimal(odds: float) -> float:
    if odds is None:
        raise ValueError("missing odds")
    odds = float(odds)
    if odds > 0:
        return 1 + odds / 100
    return 1 + 100 / abs(odds)


class EspnSource(Source):
    name = "espn"

    # Max per-cycle summary fetches (live events lacking scoreboard odds only).
    SUMMARY_CAP = 15

    def fetch_live(self, sport: str) -> list[LiveEvent]:
        path = SPORT_PATHS.get(sport)
        if not path:
            return []
        try:
            data = get_json(f"{BASE}/{path}/scoreboard", user_agent=ESPN_UA)
        except Exception:
            return []
        events: list[LiveEvent] = []
        for ev in (data.get("events") or []):
            comps = ev.get("competitions") or [{}]
            comp = comps[0]
            competitors = comp.get("competitors") or []
            names = [c.get("team", {}).get("displayName", "?") for c in competitors]
            label = " vs ".join(names) if names else ev.get("name", "?")
            state = (ev.get("status", {}).get("type", {}) or {}).get("state", "")
            markets: list[MarketPrice] = []
            for odds in comp.get("odds") or []:
                try:
                    details = odds.get("details", "")
                    mk = self._parse_market(odds, details)
                    if mk:
                        markets.append(mk)
                except Exception:
                    continue
            events.append(LiveEvent(
                event_id=str(ev.get("id", "")),
                sport=sport,
                match_label=label,
                is_live=(state == "in"),
                markets=markets,
                raw_score="",
            ))
        # Scoreboards often omit odds for live games; the per-event summary
        # endpoint still carries book prices. Backfill live events only.
        summaries = 0
        for ev in events:
            if ev.markets or not ev.is_live or summaries >= self.SUMMARY_CAP:
                continue
            extra = self._summary_markets(path, ev.event_id)
            if extra:
                ev.markets.extend(extra)
                summaries += 1
        return events

    def _summary_markets(self, path: str, event_id: str) -> list[MarketPrice]:
        """Parse DraftKings odds from the event summary endpoint."""
        try:
            data = get_json(f"{BASE}/{path}/summary?event={event_id}",
                            user_agent=ESPN_UA)
        except Exception:
            return []
        out: list[MarketPrice] = []
        for odds in data.get("pickcenter") or data.get("odds") or []:
            try:
                out.extend(self._parse_summary_odds(odds))
            except Exception:
                continue
        return out

    @staticmethod
    def _moneyline_value(node) -> float | None:
        if not isinstance(node, dict):
            return None
        for key in ("moneyLine", "moneyline", "summary", "odds", "value"):
            if node.get(key) not in (None, ""):
                try:
                    return float(node[key])
                except (TypeError, ValueError):
                    pass
        return None

    def _parse_summary_odds(self, odds: dict) -> list[MarketPrice]:
        found: list[MarketPrice] = []
        outcomes: list[OutcomePrice] = []
        home = self._moneyline_value(odds.get("homeTeamOdds"))
        away = self._moneyline_value(odds.get("awayTeamOdds"))
        draw = self._moneyline_value(odds.get("drawOdds"))
        if home is not None:
            outcomes.append(OutcomePrice(name="home", decimal_odds=_american_to_decimal(home)))
        if away is not None:
            outcomes.append(OutcomePrice(name="away", decimal_odds=_american_to_decimal(away)))
        if draw is not None:
            outcomes.append(OutcomePrice(name="draw", decimal_odds=_american_to_decimal(draw)))
        if len(outcomes) >= 2:
            found.append(MarketPrice(
                market_type="1x2" if draw is not None else "moneyline",
                line=None, outcomes=outcomes))
        ou = odds.get("overUnder")
        try:
            over = float(odds.get("overOdds")) if odds.get("overOdds") not in (None, "") else None
            under = float(odds.get("underOdds")) if odds.get("underOdds") not in (None, "") else None
        except (TypeError, ValueError):
            over = under = None
        if ou is not None and over is not None and under is not None:
            found.append(MarketPrice(market_type="total", line=str(ou), outcomes=[
                OutcomePrice(name="over", decimal_odds=_american_to_decimal(over)),
                OutcomePrice(name="under", decimal_odds=_american_to_decimal(under)),
            ]))
        return found

    def _parse_market(self, odds: dict, details: str) -> MarketPrice | None:
        outcomes: list[OutcomePrice] = []
        home_away = odds.get("homeAway") or odds.get("homeaway")
        if home_away:
            for side in ("home", "away"):
                val = (home_away.get(side) or {})
                summary = val.get("summary") or val.get("odds")
                if summary:
                    try:
                        outcomes.append(OutcomePrice(
                            name=side, decimal_odds=_american_to_decimal(summary)))
                    except (ValueError, TypeError):
                        pass
        draw = odds.get("draw") or {}
        draw_val = draw.get("summary") or draw.get("odds") if isinstance(draw, dict) else draw
        if draw_val:
            try:
                outcomes.append(OutcomePrice(
                    name="draw", decimal_odds=_american_to_decimal(draw_val)))
            except (ValueError, TypeError):
                pass
        if len(outcomes) < 2:
            return None
        market_type = "1x2" if any(o.name == "draw" for o in outcomes) else "moneyline"
        ou = odds.get("overUnder")
        line = str(ou) if ou is not None else None
        if line is not None:
            market_type = "total"
        return MarketPrice(market_type=market_type, line=line, outcomes=outcomes)

    def fetch_finals(self) -> dict[str, dict]:
        """One pass over league scoreboards: event_id -> {'home','away'} finals."""
        finals: dict[str, dict] = {}
        for path in SPORT_PATHS.values():
            try:
                data = get_json(f"{BASE}/{path}/scoreboard", user_agent=ESPN_UA)
            except Exception:
                continue
            for ev in data.get("events") or []:
                comp = (ev.get("competitions") or [{}])[0]
                state = (ev.get("status", {}).get("type", {}) or {}).get("state")
                if state != "post":
                    continue
                try:
                    paired = {c.get("homeAway"): int(c.get("score"))
                              for c in comp.get("competitors") or []}
                    finals[str(ev.get("id"))] = {"home": paired.get("home"),
                                                 "away": paired.get("away")}
                except (TypeError, ValueError):
                    continue
        return finals

    def fetch_final(self, event_id: str) -> dict | None:
        return self.fetch_finals().get(str(event_id))

File: worker/src/feeds/espn.ts
/** ESPN scoreboard client (keyless). Mirrors bot/feeds/espn.py. */
import { ESPN_UA } from "../config.ts";
import type { FinalsMap, Http, LiveEvent, MarketPrice, OutcomePrice, Source } from "./base.ts";

const BASE = "https://site.api.espn.com/apis/site/v2/sports";

export const SPORT_PATHS: Record<string, string> = {
  "soccer-epl": "soccer/eng.1",
  "soccer-laliga": "soccer/esp.1",
  "soccer-bundesliga": "soccer/ger.1",
  "soccer-seriea": "soccer/ita.1",
  "soccer-ligue1": "soccer/fra.1",
  "soccer-ucl": "soccer/uefa.champions",
  nba: "basketball/nba",
  nfl: "football/nfl",
  mlb: "baseball/mlb",
  nhl: "hockey/nhl",
  "tennis-atp": "tennis/atp",
};

export function americanToDecimal(odds: number): number {
  if (odds === null || odds === undefined || Number.isNaN(Number(odds))) {
    throw new Error("missing odds");
  }
  const o = Number(odds);
  if (o > 0) return 1 + o / 100;
  return 1 + 100 / Math.abs(o);
}

interface EspnScoreboard {
  events?: EspnEvent[];
}
interface EspnEvent {
  id?: string | number;
  name?: string;
  date?: string;
  status?: { type?: { state?: string } };
  competitions?: EspnCompetition[];
}
interface EspnCompetition {
  competitors?: Array<{ homeAway?: string; score?: string; team?: { displayName?: string } }>;
  odds?: EspnOdds[];
}
interface EspnOdds {
  details?: string;
  overUnder?: number;
  homeAway?: Record<string, { summary?: string; odds?: string | number }>;
  homeaway?: Record<string, { summary?: string; odds?: string | number }>;
  draw?: { summary?: string; odds?: string | number } | string | number;
}

function parseMarket(odds: EspnOdds): MarketPrice | null {
  const outcomes: OutcomePrice[] = [];
  const homeAway = odds.homeAway ?? odds.homeaway;
  if (homeAway) {
    for (const side of ["home", "away"]) {
      const val = homeAway[side] ?? {};
      const summary = val.summary ?? val.odds;
      if (summary !== undefined && summary !== null && summary !== "") {
        try {
          outcomes.push({ name: side, decimalOdds: americanToDecimal(Number(summary)), volume: 0 });
        } catch {
          /* skip unparseable leg */
        }
      }
    }
  }
  const draw = odds.draw;
  const drawVal =
    draw !== null && typeof draw === "object" ? (draw.summary ?? draw.odds) : (draw as string | number | undefined);
  if (drawVal !== undefined && drawVal !== null && drawVal !== "") {
    try {
      outcomes.push({ name: "draw", decimalOdds: americanToDecimal(Number(drawVal)), volume: 0 });
    } catch {
      /* skip unparseable leg */
    }
  }
  if (outcomes.length < 2) return null;
  let marketType = outcomes.some((o) => o.name === "draw") ? "1x2" : "moneyline";
  const ou = odds.overUnder;
  const line = ou !== undefined && ou !== null ? String(ou) : null;
  if (line !== null) marketType = "total";
  return { marketType, line, outcomes };
}

const SUMMARY_CAP = 15;

interface EspnSummaryOdds {
  overUnder?: number;
  spread?: number | string;
  overOdds?: number | string;
  underOdds?: number | string;
  homeTeamOdds?: Record<string, unknown>;
  awayTeamOdds?: Record<string, unknown>;
  drawOdds?: Record<string, unknown>;
}

interface EspnSummary {
  pickcenter?: EspnSummaryOdds[];
  odds?: EspnSummaryOdds[];
}

function parseSummaryOdds(odds: EspnSummaryOdds): MarketPrice[] {
  const found: MarketPrice[] = [];
  const outcomes: OutcomePrice[] = [];
  const homeML = explicitMoneyLine(odds.homeTeamOdds);
  const awayML = explicitMoneyLine(odds.awayTeamOdds);
  const drawML = explicitMoneyLine(odds.drawOdds);
  if (homeML !== null) outcomes.push({ name: "home", decimalOdds: americanToDecimal(homeML), volume: 0 });
  if (awayML !== null) outcomes.push({ name: "away", decimalOdds: americanToDecimal(awayML), volume: 0 });
  if (drawML !== null) outcomes.push({ name: "draw", decimalOdds: americanToDecimal(drawML), volume: 0 });
  if (outcomes.length >= 2) {
    found.push({
      marketType: drawML !== null ? "1x2" : "moneyline",
      line: null,
      outcomes,
    });
  }
  const ou = odds.overUnder;
  const over = odds.overOdds !== undefined && odds.overOdds !== null && odds.overOdds !== "" ? Number(odds.overOdds) : null;
  const under = odds.underOdds !== undefined && odds.underOdds !== null && odds.underOdds !== "" ? Number(odds.underOdds) : null;
  if (ou !== undefined && ou !== null && over !== null && under !== null && Number.isFinite(over) && Number.isFinite(under)) {
    found.push({
      marketType: "total",
      line: String(ou),
      outcomes: [
        { name: "over", decimalOdds: americanToDecimal(over), volume: 0 },
        { name: "under", decimalOdds: americanToDecimal(under), volume: 0 },
      ],
    });
  }
  const spread = odds.spread;
  const homeSpread = spreadValue(odds.homeTeamOdds);
  const awaySpread = spreadValue(odds.awayTeamOdds);
  if (spread !== undefined && spread !== null && homeSpread !== null && awaySpread !== null) {
    found.push({
      marketType: "handicap",
      line: String(spread),
      outcomes: [
        { name: "home", decimalOdds: americanToDecimal(homeSpread), volume: 0 },
        { name: "away", decimalOdds: americanToDecimal(awaySpread), volume: 0 },
      ],
    });
  }
  return found;
}

function explicitMoneyLine(node: unknown): number | null {
  if (!node || typeof node !== "object") return null;
  const rec = node as Record<string, unknown>;
  for (const key of ["moneyLine", "moneyline"]) {
    const v = rec[key];
    if (v !== undefined && v !== null && v !== "") {
      const n = Number(v);
      if (Number.isFinite(n)) return n;
    }
  }
  return null;
}

function spreadValue(node: unknown): number | null {
  if (!node || typeof node !== "object") return null;
  const v = (node as Record<string, unknown>)["spreadOdds"];
  if (v === undefined || v === null || v === "") return null;
  const n = Number(v);
  return Number.isFinite(n) ? n : null;
}

export class EspnSource implements Source {
  readonly name = "espn";
  async fetchLive(sport: string, http: Http): Promise<LiveEvent[]> {
    const path = SPORT_PATHS[sport];
    if (!path) return [];
    let data: EspnScoreboard;
    try {
      data = await http.getJson<EspnScoreboard>(`${BASE}/${path}/scoreboard`, undefined, ESPN_UA);
    } catch {
      return [];
    }
    const events: LiveEvent[] = [];
    for (const ev of data.events ?? []) {
      // Finished games must never be evaluated (stale odds + instant
      // settlement = bogus picks + spam loop). Mirrors bot/feeds/espn.py.
      if (ev.status?.type?.state === "post") continue;
      const comp = (ev.competitions ?? [{}])[0]!;
      const competitors = comp.competitors ?? [];
      const names = competitors.map((c) => c.team?.displayName ?? "?");
      const label = names.length > 0 ? names.join(" vs ") : (ev.name ?? "?");
      const state = ev.status?.type?.state ?? "";
      const markets: MarketPrice[] = [];
      for (const odd of comp.odds ?? []) {
        try {
          const mk = parseMarket(odd);
          if (mk) markets.push(mk);
        } catch {
          continue;
        }
      }
      events.push({
        eventId: String(ev.id ?? ""),
        sport,
        matchLabel: label,
        isLive: state === "in",
        markets,
        rawScore: "",
        startsAt: ev.date ?? null,
      });
    }
    // Scoreboards often omit odds; the per-event summary endpoint still
    // carries book prices. Backfill live events first, then pre-match
    // events starting within the next 36h. Games that started more than
    // 3h ago are stale (played yesterday) and are never backfilled.
    const now = Date.now();
    const pending: LiveEvent[] = [];
    const upcoming: LiveEvent[] = [];
    for (const ev of events) {
      if (ev.markets.length > 0) continue;
      if (ev.isLive) {
        pending.unshift(ev);
        continue;
      }
      const raw = data.events?.find((e) => String(e.id ?? "") === ev.eventId)?.date;
      if (!raw) continue;
      const kickoff = Date.parse(raw);
      if (!Number.isFinite(kickoff)) continue;
      const deltaH = (kickoff - now) / 3_600_000;
      if (deltaH >= -3 && deltaH <= 36) upcoming.push(ev);
    }
    const targets = [...pending, ...upcoming].slice(0, SUMMARY_CAP);
    await Promise.all(
      targets.map(async (ev) => {
        const extra = await this.summaryMarkets(path, ev.eventId, http);
        if (extra.length > 0) ev.markets.push(...extra);
      }),
    );
    return events;
  }

  async summaryMarkets(path: string, eventId: string, http: Http): Promise<MarketPrice[]> {
    let data: EspnSummary;
    try {
      data = await http.getJson<EspnSummary>(`${BASE}/${path}/summary`, { event: eventId }, ESPN_UA);
    } catch {
      return [];
    }
    const out: MarketPrice[] = [];
    for (const odd of data.pickcenter ?? data.odds ?? []) {
      try {
        out.push(...parseSummaryOdds(odd));
      } catch {
        continue;
      }
    }
    return out;
  }

  async fetchFinals(http: Http): Promise<FinalsMap> {
    const finals: FinalsMap = {};
    for (const path of Object.values(SPORT_PATHS)) {
      let data: EspnScoreboard;
      try {
        data = await http.getJson<EspnScoreboard>(`${BASE}/${path}/scoreboard`, undefined, ESPN_UA);
      } catch {
        continue;
      }
      for (const ev of data.events ?? []) {
        const comp = (ev.competitions ?? [{}])[0]!;
        if (ev.status?.type?.state !== "post") continue;
        try {
          const paired: Record<string, number> = {};
          for (const c of comp.competitors ?? []) {
            if (c.homeAway) paired[c.homeAway] = parseInt(c.score ?? "", 10);
          }
          if (Number.isFinite(paired["home"]) && Number.isFinite(paired["away"])) {
            const competitors = comp.competitors ?? [];
            const names = competitors.map((c) => c.team?.displayName ?? "?");
            finals[String(ev.id)] = {
              home: paired["home"]!,
              away: paired["away"]!,
              label: names.length > 0 ? names.join(" vs ") : (ev.name ?? "?"),
            };
          }
        } catch {
          continue;
        }
      }
    }
    return finals;
  }
}
